aaelf64-morello/aaelf64-morello.rst

```diff
@@ -550,6 +564,30 @@ Static Morello relocations
     |       |                         |                  | See `Call and Jump relocations`_.               |
     +-------+-------------------------+------------------+-------------------------------------------------+
 
+.. class:: aaelf64-morello-group-relocations
+
+.. table:: Group relocations to create a 16-, 32-, 48-, or 64-bit unsigned data value or address inline
+
+  +------------+--------------------------------+------------+-------------------------------------------------------------------------------------+
+  | ELF64 Code | Name                           | Operation  | Comment                                                                             |
+  +============+================================+============+=====================================================================================+
+  | 57353      | R\_MORELLO\_MOVW\_SIZE\_G0     | SIZE       | Set a MOV[KZ] immediate field to bits [15:0] of X; check that 0 <= X < 2\ :sup:`16` |
+  +------------+--------------------------------+------------+-------------------------------------------------------------------------------------+
+  | 57354      | R\_MORELLO\_MOVW\_SIZE\_G0\_NC | SIZE       | Set a MOV[KZ] immediate field to bits [15:0] of X. No overflow check                |
+  +------------+--------------------------------+------------+-------------------------------------------------------------------------------------+
+  | 57355      | R\_MORELLO\_MOVW\_SIZE\_G1     | SIZE       | Set a MOV[KZ] immediate field to bits [31:16] of X; check that 0 <= X < 2\ :sup:`32`|
+  +------------+--------------------------------+------------+-------------------------------------------------------------------------------------+
+  | 57356      | R\_MORELLO\_MOVW\_SIZE\_G1\_NC | SIZE       | Set a MOV[KZ] immediate field to bits [31:16] of X. No overflow check               |
+  +------------+--------------------------------+------------+-------------------------------------------------------------------------------------+
+  | 57357      | R\_MORELLO\_MOVW\_SIZE\_G2     | SIZE       | Set a MOV[KZ] immediate field to bits [47:32] of X; check that 0 <= X < 2\ :sup:`48`|
+  +------------+--------------------------------+------------+-------------------------------------------------------------------------------------+
+  | 57358      | R\_MORELLO\_MOVW\_SIZE\_G2\_NC | SIZE       | Set a MOV[KZ] immediate field to bits [47:32] of X. No overflow check               |
+  +------------+--------------------------------+------------+-------------------------------------------------------------------------------------+
+  | 57359      | R\_MORELLO\_MOVW\_SIZE\_G3     | SIZE       | Set a MOV[KZ] immediate field to bits [63:48] of X (no overflow check needed)       |
+  +------------+--------------------------------+------------+-------------------------------------------------------------------------------------+
```

Comment:
I think it would be nice to either mention a restriction that these relocations are only for symbols directly (and disallowed on an offset), or to specify what we do with an offset in these `size` relocations as something like "the remainder".
(My preference is to disallow with an offset).

Stock-AArch64 allows offsets on TLS accessing relocations (like `add x0, x0, #:tprel_hi12:a+48`) -- at least in the BFD linker.

If implemented naively the `size` relocations just returning the size of a symbol would mean that the pointer we end up with after a Local-Exec sequence could access outside the object we wanted and not access the start of the object.